### ISSUE TITLE
neovim api calls for autocommands

### ISSUE THREAD

**Issue by lanza** [neovim/neovim#10070]
I'm working on a plugin for debugging functionality and trying to optimize some of the code paths. One problem we have is that there doesn't seem to be a way to load a buffer to a window and then run autocommands on it as if the user opened the buffer themselves.

Given a sequence of steps
* user triggers event with window 2 selected
* plugin calls `nvim_win_set_buf` for window 1 with some other buffer

you get a raw buffer loaded into window 1 with no syntax highlighting and no autocommands ran for other plugins that might depend on it. So the work around I've used is

    target_window.buffer = new_buffer
    previous_window = vim.current.window
    vim.current.window = target_window
    vim.command("doauto SomeAutocmd")
    vim.current.window = previous_window

Clearly this is a clumsy workaround. Something along the lines of

    target_window.buffer = new_buffer
    vim.api.run_autocommand("BufEnter", target_window.buffer)

would be much nicer.

**Comment by justinmk** [comment]
> a way to load a buffer to a window and then run autocommands on it as if the user opened the buffer themselves.

`vim.command()` runs autocmds. E.g. `vim.command('edit ...')`

**Comment by bfredl** [comment]
Wouldn't `win_execute` (in vim docs but not yet implemented) solve the problem? Or I guess it could be called `nvim_win_command` in nvim API.

**Comment by bfredl** [comment]
I did a bit of testing. The issue seems to be with when a buffer is added to the buffer list and not loaded. When `nvim_win_set_buf` is called on such an unloaded buffer, the buffer is loaded without the autocmds associated with _loading_ being run, like filetype, syntax etc. When the buffer is later visited in another window using `edit {bufname}` the buffer is already considered loaded and such the autocmds is not run then either. 

I think `nvim_win_set_buf` should run the standard autocmds like `FileType` when loading an unloaded buffer. Visiting an previously unvisited buffer is per definition a side-effecty operation in vim, loading a buffer using API should not try to avoid it.

**Comment by lanza** [comment]
> I did a bit of testing. The issue seems to be with when a buffer is added to the buffer list and not loaded. When nvim_win_set_buf is called on such an unloaded buffer, the buffer is loaded without the autocmds associated with loading being run, like filetype, syntax etc. When the buffer is later visited in another window using edit {bufname} the buffer is already considered loaded and such the autocmds is not run then either.

> I think nvim_win_set_buf should run the standard autocmds like FileType when loading an unloaded buffer. Visiting an previously unvisited buffer is per definition a side-effecty operation in vim, loading a buffer using API should not try to avoid it.

Yes, the seems like a better assessment of the situation than I made. Thanks for the clarification!

"zero side effects" doesn't seem to be the right goal. It should have all the effects that follow the autocmd based lifecycle that vim normally uses. You didn't enter buffer so it shouldn't send `BufEnter` but you did read a file so it should run `FileReadCmd`.

**Comment by bfredl** [comment]
> "zero side effects" doesn't seem to be the right goal. It should have all the effects that follow the autocmd based lifecycle that vim normally uses. You didn't enter buffer so it shouldn't send BufEnter but you did read a file so it should run FileReadCmd.

Yes, that is what I meant.

**Comment by bfredl** [comment]
The same issue also exists with `nvim_open_win` with unloaded file buffer: 
```
" or any existing file with expected ft detection
let b = bufadd("CMakeLists.txt") 
call nvim_open_win(b, v:false, {'width': 50, 'height': 5, 'relative': 'editor', 'row': 5, 'col': 5, 'style': 'minimal' })
```
The filetype is not detected and loaded then either (reported by @tracyone on gitter).

This (and `nvim_win_set_buf`) should execute the `BufRead` (or `BufNewFile`) and `FileType` autocmds, but still not `BufEnter` or `WinEnter`.

**Comment by wsdjeg** [comment]
nothing need to be done with nvim_open_win, but the the ` bufadd（）` should be fixed.

**Comment by bfredl** [comment]
> nothing need to be done with nvim_open_win, but the the bufadd（） should be fixed.

What do you base this on? I don't think anything is wrong with `bufadd()`, this works well:
```
let b = bufadd("CMakeLists.txt") 
split | execute "b ".b
```
ftdetect and FileType is done when the buffer is first loaded. If this happens in `nvim_win_set_buf` or `nvim_win_open` it still need to be done.

**Comment by the42** [comment]
Instead of adding a new bug report, I guess the following (wrong?) behavior of neovim is an effect of that issue:

`:e minimal.yaml`
`vnew | r ++edit #`

According to the vim documentation, *The effect [of ++edit] is that the 'fileformat', 'fileencoding', 'bomb', etc. options are set to what has been detected for "filename". 

So if this issue is the reason for this behavior, it is a regression too. If not, I'll file a new issue.

**Comment by ddickstein** [comment]
What's the right mental model for the current semantics of `nvim_win_set_buf` re: side-effects that do and don't happen? And with the changes planned to address this issue and https://github.com/neovim/neovim/issues/13201, what will the right mental model be for its future side-effect semantics?

**Comment by justinmk** [comment]
> with the changes planned to address this issue and #13201, what will the right mental model be for its future side-effect semantics?

I think it makes sense for _all_ buffer-related API functions (except `nvim_buf_is_loaded` and maybe `nvim_buf_is_valid`) to ensure the buffer is loaded, because the _common case_ is that plugins expect that. In the performance-sensitive, uncommon case, plugins can check `nvim_buf_is_loaded` to avoid implicitly loading a buffer.

**Comment by UsamaHameed** [comment]
took a gander at this, hopefully this covers all the relevant cases #26666

**Comment by prettymuchbryce** [comment]
I recently tracked down an opaque error in a popular plugin that may be related to this issue. You can read more about it here: https://github.com/akinsho/bufferline.nvim/pull/928

Suffice to say, I wonder if this issue may extend beyond just functions on the buffer to getters/setters on the buffer such as `buflisted` as well. Maybe someone with more Neovim expertise could chime in around whether or not that's possible.

**Comment by justinmk** [comment]
I will try to revisit https://github.com/neovim/neovim/pull/26666 this week
